docs/core/extensions/httpclient-factory-keyed-di.md

```diff
@@ -0,0 +1,302 @@
+
+```c#
+services.AddHttpClient("keyed").AddAsKeyed();
+services.AddHttpClient("not-keyed");
+
+provider.GetRequiredKeyedService<HttpClient>("keyed"); // OK
+
+// Throws: No service for type 'System.Net.Http.HttpClient' has been registered.
+provider.GetRequiredKeyedService<HttpClient>("not-keyed");
+```
+
+Additionally, the Scoped lifetime of the clients can help catching cases of captive dependencies:
+
+```csharp
+services.AddHttpClient("scoped").AddAsKeyed();
+services.AddSingleton<CapturingSingleton>();
+
+// Throws: Cannot resolve scoped service 'System.Net.Http.HttpClient' from root provider.
+rootProvider.GetRequiredKeyedService<HttpClient>("scoped");
+
+using var scope = provider.CreateScope();
+scope.ServiceProvider.GetRequiredKeyedService<HttpClient>("scoped"); // OK
+
+// Throws: Cannot consume scoped service 'System.Net.Http.HttpClient' from singleton 'CapturingSingleton'.
+public class CapturingSingleton([FromKeyedServices("scoped")] HttpClient httpClient)
+//{ ...
+```
+
+## Service Lifetime Selection
+
+By default, `AddAsKeyed()` registers `HttpClient` as a Keyed _Scoped_ service. You can also explicitly specify the lifetime by passing the `ServiceLifetime` parameter to the `AddAsKeyed()` method:
+
+```csharp
+services.AddHttpClient("explicit-scoped")
+    .AddAsKeyed(ServiceLifetime.Scoped);
+
+services.AddHttpClient("singleton")
+    .AddAsKeyed(ServiceLifetime.Singleton);
+```
+
+If you call `AddAsKeyed()` within a Typed client registration, only the underlying Named client is registered as Keyed. The Typed client itself continues to be registered as a plain Transient service.
+
+### Avoid Transient HttpClient Memory Leak
+
+> [!IMPORTANT]
+> `HttpClient` is `IDisposable`, so we strongly recommend _avoiding_ Transient lifetime for Keyed `HttpClient`s.
+>
+> Registering the client as a Keyed Transient service leads to the `HttpClient` and `HttpMessageHandler` instances being _captured by DI container_, as both implement `IDisposable`. This can result in _memory leaks_ if the client is resolved multiple times within Singleton services.
+
+### Avoid Captive Dependency
+
+> [!IMPORTANT]
+> If `HttpClient` is registered either:
+>
+> - as a Keyed _Singleton_, -OR-
+> - as a Keyed _Scoped_ or _Transient_, and injected within a _long-running_ (longer than `HandlerLifetime`) application Scope, -OR-
+> - as a Keyed _Transient_, and injected into a _Singleton_ service,
+>
+> &mdash;the `HttpClient` instance becomes _captive_, and will most possibly outlive way beyond its expected `HandlerLifetime`. `HttpClientFactory` has no control over captive clients, they're NOT able to participate in the handler rotation, and it can result in [the loss of DNS changes](httpclient-factory-troubleshooting.md#httpclient-doesnt-respect-dns-changes). A similar issue [already exists](httpclient-factory.md#avoid-typed-clients-in-singleton-services) for Typed clients, which are registered as Transient services.
```

Comment:
```suggestion
> &mdash;the `HttpClient` instance becomes _captive_, and will likely outlive its expected `HandlerLifetime`. `HttpClientFactory` has no control over captive clients, they're NOT able to participate in the handler rotation, and it can result in [the loss of DNS changes](httpclient-factory-troubleshooting.md#httpclient-doesnt-respect-dns-changes). A similar issue [already exists](httpclient-factory.md#avoid-typed-clients-in-singleton-services) for Typed clients, which are registered as Transient services.
```